### ISSUE TITLE
KeyStore JKS 

### ISSUE THREAD

**Issue by junioorosa** [zozfabio/emissor-gnre-httpclient#1]
Como eu consigo o KeyStore JKS com o certificado do WebService?

**Comment by zozfabio** [comment]
Basicamente você deve criar um novo keystore no formato JKS, depois importar os certificados dos web services e salvar o arquivo. Depois você só configura o caminho do arquivo e a senha que você definiu para o keystore na classe ClientFactory.

Eu costumo utilizar uma ferramenta chamada Portecle para manipular keystores.

Para criar um novo keystore [aqui](https://portecle.sourceforge.net/create-keystore.html) tem instruções e para importar o certificado de um site ou webservice você pode seguir [estas instruções](https://portecle.sourceforge.net/examine-tls.html).

[Neste artigo](https://confluence.atlassian.com/kb/how-to-import-a-public-ssl-certificate-into-a-jvm-867025849.html) tem mais instruções passo a passo de como importar o certificado de um site em um keystore existente.

**Comment by junioorosa** [comment]
Segui todos os passos que vc destacou, conseguindo assim gerar a keystore no formato JKS, porém ao executar não tive sucesso.

Para o Web Service https://www.testegnre.pe.gov.br/gnreWS/services/GnreConfigUF com o model:
```
<?xml version="1.0" encoding="UTF-8"?>
<soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:gnr="http://www.gnre.pe.gov.br/webservice/GnreConfigUF">
   <soap:Header>
      <gnr:gnreCabecMsg>
         <gnr:versaoDados>1.00</gnr:versaoDados>
      </gnr:gnreCabecMsg>
   </soap:Header>
   <soap:Body>
      <gnr:gnreDadosMsg>
         <TConsultaConfigUf xmlns="http://www.gnre.pe.gov.br">
            <ambiente>2</ambiente>
            <uf>MG</uf>
            <receita>100048</receita>
         </TConsultaConfigUf>
      </gnr:gnreDadosMsg>
   </soap:Body>
</soap:Envelope>
```
Está me gerando o seguinte erro:
```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>403 Forbidden</title>
</head><body>
<h1>Forbidden</h1>
<p>You don't have permission to access this resource.</p>
<hr>
<address>IBM_HTTP_Server at www.testegnre.pe.gov.br Port 443</address>
</body></html>
```

Percebi que se eu coloco no endereço do Web Service a referência para a porta 444 me retorna outro erro:
https://www.testegnre.pe.gov.br:444/gnreWS/services/GnreConfigUF
```
<?xml version="1.0" encoding="UTF-8"?>
<soapenv:Envelope xmlns:soapenv="http://www.w3.org/2003/05/soap-envelope" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <soapenv:Body>
      <gnreRespostaMsg xmlns="http://www.gnre.pe.gov.br/webservice/GnreConfigUF">
         <ns1:TConfigUf xmlns:ns1="http://www.gnre.pe.gov.br" xmlns="http://www.gnre.pe.gov.br">
            <ns1:ambiente>2</ns1:ambiente>
            <ns1:uf>MG</ns1:uf>
            <ns1:situacaoConsulta>
               <ns1:codigo>199</ns1:codigo>
               <ns1:descricao>Erro ao validar comunicação: null</ns1:descricao>
            </ns1:situacaoConsulta>
         </ns1:TConfigUf>
      </gnreRespostaMsg>
   </soapenv:Body>
</soapenv:Envelope>
```

Obs.: 
- Já solicitei o uso dos Web Services no Portal GNRE, tanto homologação quanto produção.
- Meu certificado está valido. Também verifiquei se o CNPJ que solicitei o acesso era o mesmo do certificado.

Segue a configuração da classe ClientFactory
![image](https://github.com/zozfabio/emissor-gnre-httpclient/assets/86578051/9a92e2d9-663c-4e67-937b-32be587d3666)

Se puder ajudar eu agradeço, pois parece que estou rodando em círculos por conta dessa implementação.

**Comment by zozfabio** [comment]
Sim, entendo amigo, passei um sufoco também. infelizmente faz tempo que eu implementei a integração e eu não lembro de muita coisa
acessando o site, parece que o serviço de teste está realmente rodando na porta 444 agora, o certificado que você importou no keystore também foi da porta 444? você pode importar o certificados da porta 443 e 444 no mesmo keystore só para garantir, inclusive do webservice de produção também pode ser no mesmo keystore.
você já verificou se esse código de erro consta no manual em pdf? lá tem documentados vários códigos de erros para depuração.
a solicitação de acesso foi solicitado nesse formulário?
![image](https://github.com/zozfabio/emissor-gnre-httpclient/assets/3035626/5a66e30a-f337-44c6-974d-04efc5007d6a)
porque tem que ser feito alí para conseguir acesso ao serviço de teste que eu me lembro.
outra coisa, essa tela abaixo me ajudou um pouco para conseguir passar os dados corretos, estado, receita, etc.
![image](https://github.com/zozfabio/emissor-gnre-httpclient/assets/3035626/1b2e7552-ea42-43ac-b7a7-523d378a4e2c)
no momento eu não posso testar nada porque não tenho um acesso e certificado ao webservice, então estou só na base do chute.
